File: functions/lib/event-duplicate-review.js
/**
 * Detects transferable-field changes on events marked as duplicates.
 * Stores a baseline at link time and flags field groups for moderator review.
 */

const FIELD_GROUPS = Object.freeze([
  "photos",
  "linkedSpots",
  "title",
  "description",
  "location",
  "schedule",
  "website",
]);

/**
 * @param {*} value
 * @return {string|null}
 */
function toNonEmptyString(value) {
  if (typeof value !== "string") return null;
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : null;
}

/**
 * @param {*} value
 * @return {Date|null}
 */
function normalizeDate(value) {
  if (!value) return null;
  if (value instanceof Date && !Number.isNaN(value.getTime())) return value;
  if (typeof value.toDate === "function") {
    const asDate = value.toDate();
    if (asDate instanceof Date && !Number.isNaN(asDate.getTime())) {
      return asDate;
    }
  }
  if (typeof value === "string") {
    const parsed = new Date(value);
    if (!Number.isNaN(parsed.getTime())) return parsed;
  }
  return null;
}

/**
 * @param {*} value
 * @return {number|null}
 */
function normalizeNumber(value) {
  if (value == null || value === "") return null;
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

/**
 * @param {*} value
 * @return {string[]}
 */
function normalizeStringList(value) {
  if (!Array.isArray(value)) return [];
  const result = [];
  for (const item of value) {
    const normalized = toNonEmptyString(item);
    if (normalized) result.push(normalized);
  }
  return result;
}

/**
 * @param {Object|null|undefined} data
 * @return {string|null}
 */
function getDuplicateOf(data) {
  if (!data || typeof data !== "object") return null;
  return toNonEmptyString(data.duplicateOf);
}

/**
 * @param {Object|null|undefined} data
 * @return {boolean}
 */
function hasReviewBaseline(data) {
  if (!data || typeof data !== "object") return false;
  const baseline = data.duplicateReviewBaseline;
  return Boolean(baseline) &&
    typeof baseline === "object" &&
    !Array.isArray(baseline);
}

/**
 * Snapshot of transferable fields used as the last-reviewed baseline.
 * @param {Object|null|undefined} data
 * @return {Object}
 */
function buildDuplicateReviewSnapshot(data) {
  const source = data && typeof data === "object" ? data : {};
  return {
    title: toNonEmptyString(source.title) || "",
    description: toNonEmptyString(source.description),
    websiteUrl: toNonEmptyString(source.websiteUrl),
    imageUrls: normalizeStringList(source.imageUrls),
    spotIds: normalizeStringList(source.spotIds),
    spotListIds: normalizeStringList(source.spotListIds),
    latitude: normalizeNumber(source.latitude),
    longitude: normalizeNumber(source.longitude),
    address: toNonEmptyString(source.address),
    city: toNonEmptyString(source.city),
    countryCode: toNonEmptyString(source.countryCode),
    startAt: normalizeDate(source.startAt),
    endAt: normalizeDate(source.endAt),
    isDateOnly: source.isDateOnly === true,
    timeZone: toNonEmptyString(source.timeZone),
    timeZoneSource: toNonEmptyString(source.timeZoneSource),
  };
}

/**
 * @param {string[]} left
 * @param {string[]} right
 * @return {boolean}
 */
function stringListsEqual(left, right) {
  if (left.length !== right.length) return false;
  return left.every((value, index) => value === right[index]);
}

/**
 * @param {string[]} left
 * @param {string[]} right
 * @return {boolean}
 */
function stringSetsEqual(left, right) {
  if (left.length !== right.length) return false;
  const sortedLeft = [...left].sort();
  const sortedRight = [...right].sort();
  return sortedLeft.every((value, index) => value === sortedRight[index]);
}

/**
 * @param {Date|null} left
 * @param {Date|null} right
 * @return {boolean}
 */
function datesEqual(left, right) {
  if (left == null && right == null) return true;
  if (left == null || right == null) return false;
  return left.getTime() === right.getTime();
}

/**
 * @param {number|null} left
 * @param {number|null} right
 * @return {boolean}
 */
function numbersEqual(left, right) {
  if (left == null && right == null) return true;
  if (left == null || right == null) return false;
  return left === right;
}

/**
 * @param {Object} baseline
 * @param {Object} current
 * @return {string[]}
 */
function diffDuplicateFieldGroups(baseline, current) {
  const left = buildDuplicateReviewSnapshot(baseline);
  const right = buildDuplicateReviewSnapshot(current);
  const changed = [];

  if (!stringListsEqual(left.imageUrls, right.imageUrls)) {
    changed.push("photos");
  }
  if (
    !stringSetsEqual(left.spotIds, right.spotIds) ||
    !stringSetsEqual(left.spotListIds, right.spotListIds)
  ) {
    changed.push("linkedSpots");
  }
  if (left.title !== right.title) changed.push("title");
  if (left.description !== right.description) changed.push("description");
  if (
    !numbersEqual(left.latitude, right.latitude) ||
    !numbersEqual(left.longitude, right.longitude) ||
    left.address !== right.address ||
    left.city !== right.city ||
    left.countryCode !== right.countryCode
  ) {
    changed.push("location");
  }
  if (
    !datesEqual(left.startAt, right.startAt) ||
    !datesEqual(left.endAt, right.endAt) ||
    left.isDateOnly !== right.isDateOnly ||
    left.timeZone !== right.timeZone ||
    left.timeZoneSource !== right.timeZoneSource
  ) {
    changed.push("schedule");
  }
  if (left.websiteUrl !== right.websiteUrl) changed.push("website");

  return changed;
}

/**
 * @param {*} value
 * @return {string[]}
 */
function normalizeChangedFields(value) {
  if (!Array.isArray(value)) return [];
  const allowed = new Set(FIELD_GROUPS);
  const seen = new Set();
  const result = [];
  for (const item of value) {
    const key = toNonEmptyString(item);
    if (!key || !allowed.has(key) || seen.has(key)) continue;
    seen.add(key);
    result.push(key);
  }
  return result;
}

/**
 * @param {string[]} left
 * @param {string[]} right
 * @return {boolean}
 */
function changedFieldsEqual(left, right) {
  return stringListsEqual(
      [...normalizeChangedFields(left)].sort(),
      [...normalizeChangedFields(right)].sort(),
  );
}

/**
 * @param {Object|null|undefined} data
 * @return {boolean}
 */
function hasStoredReviewFields(data) {
  if (!data || typeof data !== "object") return false;
  if (hasReviewBaseline(data)) return true;
  if (data.duplicateHasPendingChanges === true) return true;
  if (data.duplicateHasPendingChanges === false) return true;
  return normalizeChangedFields(data.duplicateChangedFields).length > 0;
}

/**
 * @param {Object|null|undefined} data
 * @param {string[]} changedFields
 * @param {boolean} expectBaseline
 * @return {boolean}
 */
function pendingStateMatches(data, changedFields, expectBaseline) {
  const pending = changedFields.length > 0;
  const storedPending = data && data.duplicateHasPendingChanges === true;
  if (storedPending !== pending) return false;
  if (!changedFieldsEqual(data && data.duplicateChangedFields, changedFields)) {
    return false;
  }
  if (expectBaseline && !hasReviewBaseline(data)) return false;
  return true;
}

/**
 * Builds a Firestore update for duplicate review state, or null when no write
 * is needed (keeps onEventWritten idempotent).
 * @param {Object|null|undefined} beforeData
 * @param {Object|null|undefined} afterData
 * @param {*} fieldValueDelete FieldValue.delete() sentinel
 * @return {Object|null}
 */
function buildDuplicateReviewUpdate(beforeData, afterData, fieldValueDelete) {
  const afterDup = getDuplicateOf(afterData);
  if (!afterDup) {
    if (!hasStoredReviewFields(afterData)) return null;
    return {
      duplicateReviewBaseline: fieldValueDelete,
      duplicateChangedFields: fieldValueDelete,
      duplicateHasPendingChanges: fieldValueDelete,
    };
  }

  const beforeDup = getDuplicateOf(beforeData);
  const newlyLinked = !beforeDup && Boolean(afterDup);
  const afterSnap = buildDuplicateReviewSnapshot(afterData);

  if (newlyLinked) {
    if (
      pendingStateMatches(afterData, [], true) &&
      diffDuplicateFieldGroups(
          afterData.duplicateReviewBaseline,
          afterSnap,
      ).length === 0
    ) {
      return null;
    }
    return {
      duplicateReviewBaseline: afterSnap,
      duplicateChangedFields: fieldValueDelete,
      duplicateHasPendingChanges: false,
    };
  }

  if (!hasReviewBaseline(afterData)) {
    const beforeSnap = beforeData ?
      buildDuplicateReviewSnapshot(beforeData) :
      afterSnap;
    const changed = beforeData ?
      diffDuplicateFieldGroups(beforeSnap, afterSnap) :
      [];
    if (changed.length > 0) {
      if (pendingStateMatches(afterData, changed, true)) return null;
      return {
        duplicateReviewBaseline: beforeSnap,
        duplicateChangedFields: changed,
        duplicateHasPendingChanges: true,
      };
    }
    if (pendingStateMatches(afterData, [], true)) return null;
    return {
      duplicateReviewBaseline: afterSnap,
      duplicateChangedFields: fieldValueDelete,
      duplicateHasPendingChanges: false,
    };
  }

  const changed = diffDuplicateFieldGroups(
      afterData.duplicateReviewBaseline,
      afterSnap,
  );
  if (pendingStateMatches(afterData, changed, true)) return null;
  if (changed.length > 0) {
    return {
      duplicateChangedFields: changed,
      duplicateHasPendingChanges: true,
    };
  }
  return {
    duplicateChangedFields: fieldValueDelete,
    duplicateHasPendingChanges: false,
  };
}

module.exports = {
  FIELD_GROUPS,
  buildDuplicateReviewSnapshot,
  buildDuplicateReviewUpdate,
  diffDuplicateFieldGroups,
  getDuplicateOf,
};

File: lib/utils/event_duplicate_review.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/material.dart';

import '../l10n/app_localizations.dart';
import '../models/parkour_event.dart';

/// Transferable field groups used when reviewing post-link duplicate changes.
enum EventDuplicateFieldGroup {
  photos,
  linkedSpots,
  title,
  description,
  location,
  schedule,
  website,
}

extension EventDuplicateFieldGroupX on EventDuplicateFieldGroup {
  String get firestoreValue {
    switch (this) {
      case EventDuplicateFieldGroup.photos:
        return 'photos';
      case EventDuplicateFieldGroup.linkedSpots:
        return 'linkedSpots';
      case EventDuplicateFieldGroup.title:
        return 'title';
      case EventDuplicateFieldGroup.description:
        return 'description';
      case EventDuplicateFieldGroup.location:
        return 'location';
      case EventDuplicateFieldGroup.schedule:
        return 'schedule';
      case EventDuplicateFieldGroup.website:
        return 'website';
    }
  }

  String label(AppLocalizations l10n) {
    switch (this) {
      case EventDuplicateFieldGroup.photos:
        return l10n.eventDetailMarkDuplicatePhotos;
      case EventDuplicateFieldGroup.linkedSpots:
        return l10n.eventDetailMarkDuplicateLinkedSpots;
      case EventDuplicateFieldGroup.title:
        return l10n.eventDetailMarkDuplicateEventTitle;
      case EventDuplicateFieldGroup.description:
        return l10n.eventDetailMarkDuplicateDescription;
      case EventDuplicateFieldGroup.location:
        return l10n.eventDetailMarkDuplicateLocation;
      case EventDuplicateFieldGroup.schedule:
        return l10n.eventDetailMarkDuplicateSchedule;
      case EventDuplicateFieldGroup.website:
        return l10n.eventDetailMarkDuplicateWebsite;
    }
  }
}

EventDuplicateFieldGroup? eventDuplicateFieldGroupFromString(String value) {
  switch (value.trim()) {
    case 'photos':
      return EventDuplicateFieldGroup.photos;
    case 'linkedSpots':
      return EventDuplicateFieldGroup.linkedSpots;
    case 'title':
      return EventDuplicateFieldGroup.title;
    case 'description':
      return EventDuplicateFieldGroup.description;
    case 'location':
      return EventDuplicateFieldGroup.location;
    case 'schedule':
      return EventDuplicateFieldGroup.schedule;
    case 'website':
      return EventDuplicateFieldGroup.website;
    default:
      return null;
  }
}

List<EventDuplicateFieldGroup> parseDuplicateChangedFieldGroups(
  Iterable<String> values,
) {
  final result = <EventDuplicateFieldGroup>[];
  final seen = <EventDuplicateFieldGroup>{};
  for (final value in values) {
    final group = eventDuplicateFieldGroupFromString(value);
    if (group == null || seen.contains(group)) continue;
    seen.add(group);
    result.add(group);
  }
  return result;
}

bool _nullableStringsEqual(String? left, String? right) {
  final normalizedLeft = left?.trim();
  final normalizedRight = right?.trim();
  final leftValue = (normalizedLeft == null || normalizedLeft.isEmpty)
      ? null
      : normalizedLeft;
  final rightValue = (normalizedRight == null || normalizedRight.isEmpty)
      ? null
      : normalizedRight;
  return leftValue == rightValue;
}

bool _stringListsEqual(List<String> left, List<String> right) {
  if (left.length != right.length) return false;
  for (var i = 0; i < left.length; i++) {
    if (left[i] != right[i]) return false;
  }
  return true;
}

bool _stringSetsEqual(List<String> left, List<String> right) {
  if (left.length != right.length) return false;
  final sortedLeft = [...left]..sort();
  final sortedRight = [...right]..sort();
  return _stringListsEqual(sortedLeft, sortedRight);
}

/// Field groups that differ between [previous] (baseline) and [current].
List<EventDuplicateFieldGroup> changedEventDuplicateFieldGroups({
  required ParkourEvent previous,
  required ParkourEvent current,
}) {
  final changed = <EventDuplicateFieldGroup>[];
  if (!_stringListsEqual(previous.imageUrls, current.imageUrls)) {
    changed.add(EventDuplicateFieldGroup.photos);
  }
  if (!_stringSetsEqual(previous.spotIds, current.spotIds) ||
      !_stringSetsEqual(previous.spotListIds, current.spotListIds)) {
    changed.add(EventDuplicateFieldGroup.linkedSpots);
  }
  if (!_nullableStringsEqual(previous.title, current.title)) {
    changed.add(EventDuplicateFieldGroup.title);
  }
  if (!_nullableStringsEqual(previous.description, current.description)) {
    changed.add(EventDuplicateFieldGroup.description);
  }
  if (previous.latitude != current.latitude ||
      previous.longitude != current.longitude ||
      !_nullableStringsEqual(previous.address, current.address) ||
      !_nullableStringsEqual(previous.city, current.city) ||
      !_nullableStringsEqual(previous.countryCode, current.countryCode)) {
    changed.add(EventDuplicateFieldGroup.location);
  }
  if (previous.startAt.toUtc() != current.startAt.toUtc() ||
      previous.endAt?.toUtc() != current.endAt?.toUtc() ||
      previous.isDateOnly != current.isDateOnly ||
      !_nullableStringsEqual(previous.timeZone, current.timeZone) ||
      !_nullableStringsEqual(previous.timeZoneSource, current.timeZoneSource)) {
    changed.add(EventDuplicateFieldGroup.schedule);
  }
  if (!_nullableStringsEqual(previous.websiteUrl, current.websiteUrl)) {
    changed.add(EventDuplicateFieldGroup.website);
  }
  return changed;
}

/// Snapshot stored on the duplicate at last review / mark-as-duplicate.
Map<String, dynamic> buildEventDuplicateReviewBaseline(ParkourEvent event) {
  return {
    'title': event.title,
    if (event.description != null && event.description!.trim().isNotEmpty)
      'description': event.description!.trim(),
    if (event.websiteUrl != null && event.websiteUrl!.trim().isNotEmpty)
      'websiteUrl': event.websiteUrl!.trim(),
    'imageUrls': List<String>.from(event.imageUrls),
    'spotIds': List<String>.from(event.spotIds),
    'spotListIds': List<String>.from(event.spotListIds),
    if (event.latitude != null) 'latitude': event.latitude,
    if (event.longitude != null) 'longitude': event.longitude,
    if (event.address != null && event.address!.trim().isNotEmpty)
      'address': event.address!.trim(),
    if (event.city != null && event.city!.trim().isNotEmpty)
      'city': event.city!.trim(),
    if (event.countryCode != null && event.countryCode!.trim().isNotEmpty)
      'countryCode': event.countryCode!.trim(),
    'startAt': Timestamp.fromDate(event.startAt.toUtc()),
    if (event.endAt != null) 'endAt': Timestamp.fromDate(event.endAt!.toUtc()),
    'isDateOnly': event.isDateOnly,
    if (event.timeZone != null && event.timeZone!.trim().isNotEmpty)
      'timeZone': event.timeZone!.trim(),
    if (event.timeZoneSource != null && event.timeZoneSource!.trim().isNotEmpty)
      'timeZoneSource': event.timeZoneSource!.trim(),
  };
}

/// Clears pending-change flags and rewrites the baseline to current fields.
Map<String, dynamic> buildDuplicateReviewAcknowledgedUpdates(
  ParkourEvent event,
) {
  return {
    'duplicateReviewBaseline': buildEventDuplicateReviewBaseline(event),
    'duplicateChangedFields': FieldValue.delete(),
    'duplicateHasPendingChanges': false,
    'updatedAt': FieldValue.serverTimestamp(),
  };
}

/// Deletes review fields together with duplicate status.
Map<String, dynamic> buildDuplicateReviewClearUpdates() {
  return {
    'duplicateReviewBaseline': FieldValue.delete(),
    'duplicateChangedFields': FieldValue.delete(),
    'duplicateHasPendingChanges': FieldValue.delete(),
  };
}

String formatEventDuplicateFieldGroupValue({
  required BuildContext context,
  required ParkourEvent event,
  required EventDuplicateFieldGroup group,
  required AppLocalizations l10n,
}) {
  switch (group) {
    case EventDuplicateFieldGroup.photos:
      return l10n.eventDuplicateChangesPhotosValue(event.imageUrls.length);
    case EventDuplicateFieldGroup.linkedSpots:
      return l10n.eventDuplicateChangesLinkedSpotsValue(
        event.spotIds.length + event.spotListIds.length,
      );
    case EventDuplicateFieldGroup.title:
      final title = event.title.trim();
      return title.isEmpty ? l10n.eventDuplicateChangesNoValue : title;
    case EventDuplicateFieldGroup.description:
      final description = event.description?.trim() ?? '';
      if (description.isEmpty) return l10n.eventDuplicateChangesNoValue;
      if (description.length <= 80) return description;
      return '${description.substring(0, 79)}…';
    case EventDuplicateFieldGroup.location:
      final city = event.city?.trim();
      if (city != null && city.isNotEmpty) return city;
      final address = event.address?.trim();
      if (address != null && address.isNotEmpty) return address;
      if (event.latitude != null && event.longitude != null) {
        return '${event.latitude!.toStringAsFixed(5)}, '
            '${event.longitude!.toStringAsFixed(5)}';
      }
      return l10n.eventDuplicateChangesNoValue;
    case EventDuplicateFieldGroup.schedule:
      final material = MaterialLocalizations.of(context);
      final start = event.startAt.toLocal();
      final date = material.formatMediumDate(start);
      if (event.isDateOnly) return date;
      final time = material.formatTimeOfDay(TimeOfDay.fromDateTime(start));
      return '$date $time';
    case EventDuplicateFieldGroup.website:
      final website = event.websiteUrl?.trim() ?? '';
      return website.isEmpty ? l10n.eventDuplicateChangesNoValue : website;
  }
}
